x/meshsecurity/keeper/keeper.go

```diff
@@ -0,0 +1,107 @@
+	// should be the x/gov module account.
+	authority string
+}
+
+// NewKeeper constructor with vanilla sdk keepers
+func NewKeeper(cdc codec.Codec, storeKey storetypes.StoreKey, account types.AccountKeeper, bank types.SDKBankKeeper, staking types.SDKStakingKeeper, authority string) *Keeper {
+	return NewKeeperX(cdc, storeKey, account, NewBankKeeperAdapter(bank), NewStakingKeeperAdapter(staking, bank), authority)
+}
+
+// NewKeeperX constructor with extended Osmosis SDK keepers
+func NewKeeperX(cdc codec.Codec, storeKey storetypes.StoreKey, accountKeeper types.AccountKeeper, bank types.XBankKeeper, staking types.XStakingKeeper, authority string) *Keeper {
+	return &Keeper{storeKey: storeKey, cdc: cdc, accountKeeper: accountKeeper, bank: bank, staking: staking, authority: authority}
+}
+
+func (k Keeper) IsAuthorized(ctx sdk.Context, contractAddr sdk.AccAddress) (bool, error) {
+	// TODO implement me
+	panic("implement me")
+}
+
+// GetParams returns the total set of wasm parameters.
+func (k Keeper) GetParams(ctx sdk.Context) types.Params {
+	var params types.Params
+	store := ctx.KVStore(k.storeKey)
+	bz := store.Get(types.ParamsKey)
+	if bz == nil {
+		return params
+	}
+
+	// todo: impl
+	//	k.cdc.MustUnmarshal(bz, &params)
+	return params
+}
+
+// SetParams sets all wasm parameters.
+func (k Keeper) SetParams(ctx sdk.Context, ps types.Params) error {
+	if err := ps.ValidateBasic(); err != nil {
+		return err
+	}
+	// todo: impl
+	//store := ctx.KVStore(k.storeKey)
+	//bz, err := k.cdc.Marshal(&ps)
+	//if err != nil {
+	//	return err
+	//}
+	//store.Set(types.ParamsKey, bz)
+	return nil
+}
+
+// getMaxCapLimit the cap limit is set per consumer contract. Different providers can have different limits
+func (k Keeper) getMaxCapLimit(ctx sdk.Context, actor sdk.AccAddress) math.Int {
+	return k.mustLoadInt(ctx, k.storeKey, types.BuildMaxCapLimitKey(actor))
+}
+
+// getTotalDelegatedAmount returns the total amount delegated by the given consumer contract
+func (k Keeper) getTotalDelegatedAmount(ctx sdk.Context, actor sdk.AccAddress) math.Int {
+	return k.mustLoadInt(ctx, k.storeKey, types.BuildTotalDelegatedAmountKey(actor))
+}
+
+// helper to deserialize a math.Int from store.
+// Panics when Unmarshal fails
+func (k Keeper) mustLoadInt(ctx sdk.Context, storeKey storetypes.StoreKey, key []byte) math.Int {
+	store := ctx.KVStore(storeKey)
+	bz := store.Get(key)
+	if bz == nil {
+		return sdk.ZeroInt()
+	}
+	var r math.Int
+	if err := r.Unmarshal(bz); err != nil {
+		panic(err)
+	}
+	return r
+}
+
+func (k Keeper) setTotalDelegatedAmount(ctx sdk.Context, actor sdk.AccAddress, newAmount math.Int) {
+	store := ctx.KVStore(k.storeKey)
+	bz, err := newAmount.Marshal()
+	if err != nil { // always nil
+		panic(err)
+	}
+	store.Set(types.BuildTotalDelegatedAmountKey(actor), bz)
+}
+
+func (k Keeper) getOrCreateIntermediaryAccount(ctx sdk.Context, actor sdk.AccAddress, addr sdk.ValAddress) sdk.AccAddress {
```

Comment:
Let's discuss the need of intermediary accounts.

In my mind, if the caller had "max_cap" set, and thus could execute delegate, the tokens would be minted and delegated *from it's account*. This would mean it could later redelegate or withdraw rewards as normal.

It also has the downside, it could call the standard `StakingMsg::Undelegate` command and thus get these tokens liquid. I think that is very unlikely as such a contract would be reviewed and permissioned by governance and such an obvious attack would be found (and any other bug can mint millions of tokens and stake to the wrong validator, so the contract must be reviewed well in any case).

If we really want to isolate the contract from these virtual tokens (which is not a bad design, just a bit overkill IMO), we must minimally add a "withdraw rewards" to the CustomMsg.VirtualStake

Reply:
Good points! I copied this from superfluid staking but their case is a bit different than ours as we have an address per contract. Intermediary accounts add a lot of complexity and I would be very happy to remove them.
Some additional safety net to prevent delegate/undelegate from a contract makes sense. I would include all staking and stargate messages.

Btw, `Redelegating` should be completely avoided by the contract. The SDK has limits on numbers and rejects transitive redelegation. Redelegations are also affected on slashing. It would be a nightmare to keep track of this.